tools/apiview/parsers/cpp-api-parser/ApiViewProcessor/JsonDumper.hpp

```diff
@@ -0,0 +1,313 @@
+    Comment = 10,
+    DocumentRangeStart = 11,
+    DocumentRangeEnd = 12,
+    DeprecatedRangeStart = 13,
+    DeprecatedRangeEnd = 14,
+    SkipDiffRangeStart = 15,
+    SkipDiffRangeEnd = 16,
+    InheritanceInfoStart = 17,
+    InheritanceInfoEnd = 18
+  };
+
+public:
+  JsonDumper(
+      std::string_view reviewName,
+      std::string_view serviceName,
+      std::string_view packageName,
+      std::string_view packageVersion = "")
+      : AstDumper()
+  {
+    m_json["Name"] = reviewName;
+    m_json["Language"] = "C++";
+    m_json["ServiceName"] = serviceName;
+    m_json["PackageName"] = packageName;
+    if (!packageVersion.empty())
+    {
+      m_json["PackageVersion"] = packageVersion;
+    }
+    m_json["Tokens"] = nlohmann::json::array();
+  }
+
+  void DumpToFile(std::ostream& outfile) { outfile << m_json; }
+  nlohmann::json const& GetJson() { return m_json; }
+
+  // Each ApiView node has 4 mandatory members:
+  //
+  // DefinitionId: ID used in the Navigation pane for type navigation.
+  // NavigateToId: ???
+  // Value: Value to display in ApiView (mandatory)
+  // Kind: Type of node, used for color coding output.
+
+  virtual void InsertWhitespace(int count) override
+  {
+    std::string whiteSpace(count, ' ');
+    m_json["Tokens"].push_back(
+        {{"DefinitionId", nullptr},
+         {"NavigateToId", nullptr},
+         {"Value", whiteSpace},
+         {"Kind", TokenKinds::Whitespace}});
+  }
+  virtual void InsertNewline() override
+  {
+    m_json["Tokens"].push_back(
+        {{"DefinitionId", nullptr},
+         {"NavigateToId", nullptr},
+         {"Value", nullptr},
+         {"Kind", TokenKinds::Newline}});
+  }
+  virtual void InsertKeyword(std::string_view const& keyword) override
+  {
+    m_json["Tokens"].push_back(
+        {{"DefinitionId", nullptr},
+         {"NavigateToId", nullptr},
+         {"Value", keyword},
+         {"Kind", TokenKinds::Keyword}});
+  }
+  virtual void InsertText(std::string_view const& text) override
+  {
+    m_json["Tokens"].push_back(
+        {{"DefinitionId", nullptr},
+         {"NavigateToId", nullptr},
+         {"Value", text},
+         {"Kind", TokenKinds::Text}});
+  }
+  virtual void InsertPunctuation(const char punctuation) override
+  {
+    std::string punctuationString{punctuation};
+    m_json["Tokens"].push_back(
+        {{"DefinitionId", nullptr},
+         {"NavigateToId", nullptr},
+         {"Value", punctuationString},
+         {"Kind", TokenKinds::Punctuation}});
+  }
+  virtual void InsertLineIdMarker() override
+  {
+    m_json["Tokens"].push_back(
+        {{"DefinitionId", nullptr},
+         {"NavigateToId", nullptr},
+         {"Value", nullptr},
+         {"Kind", TokenKinds::LineIdMarker}}); // Not clear if this is used at all.
+  }
+  virtual void InsertTypeName(std::string_view const& type, std::string_view const& navigationId)
+      override
+  {
+    m_json["Tokens"].push_back(
+        {{"DefinitionId", navigationId},
```

Comment:
DefinitionID == navigationID?  I don't believe those are a 1:1 mapping

Reply:
I've been trying to find out what "NavigateToId" means from Praveen, it's one of my outstanding questions. For now, I just set the two to the same value.